### DISCUSSION TITLE
__dirname pointing to `/` (during dev)

### DISCUSSION THREAD

**Discussion by remy** [vercel/next.js#14341] [Help] [answered]
I'm trying to port some code to Next, and I'd say I was pretty familiar with how everything works, but this has stumped me (and there doesn't seem to be a mention in these discussions) - so:

During dev, `__dirname` is pointing to `/` - when my project, let alone the actual directory, is much deeper than that (more like `/Users/remy/dev/myapp/lib/emails/`).

I'm trying to send emails from an API route, and I load the HTML templates from files in the `__dirname` directory (which worked when I was running this under Express). But running from an `/api/get-token` route (which sends an email), the directory name isn't right so I can't load up the static files to send the emails.

Any ideas - or alternative suggestions?

**Comment by jamesmosier** [comment, marked as answer]
What does `process.cwd()` point to?

I remember [seeing some docs](https://nextjs.org/docs/basic-features/data-fetching#reading-files-use-processcwd) about this (related to `getStaticProps` at least)

**Comment by timneutkens** [comment]
When code is compiled it ends up in another directory structure (for example `.next/server/pages/index.js`) and the Node.js `__dirname` would always be incorrect. Webpack automatically replaces `__dirname` in order to provide "compatibility", basically just not breaking on the code itself. Especially between development and production there could be a big difference given that if `__dirname` was supported it would have to be a hardcoded absolute path, meaning the generated bundles would only be able to run on the machine they were built on. Hence why `process.cwd()` is recommended, which gives you the working directory of where the command was executed instead

**Reply by eivindml** [reply]
This is all very confusing. Why does this example use `__dirname`? https://vercel.com/support/articles/how-can-i-use-files-in-serverless-functions The example won't work by the way. I have disabled web pack 5, without any luck.

It works locally with `process.cwd()` but not on Vercel.

**Reply by Babatunde50** [reply]
I am having issues getting things to work on vercel too.

> This is all very confusing. Why does this example use `__dirname`? https://vercel.com/support/articles/how-can-i-use-files-in-serverless-functions The example won't work by the way. I have disabled web pack 5, without any luck.
> 
> It works locally with `process.cwd()` but not on Vercel.

**Reply by chrisui** [reply]
How does vercel determine which directories & files to package up and push to lambda?

I have noticed this not working for me now but ofc public is a folder outside .next which works so I presume there is a config somewhere? Unfortunately there are private files I wish to hide so placing in public is not an option

**Reply by chrisui** [reply]
Found https://vercel.com/docs/runtimes#advanced-usage/technical-details/including-additional-files

Now just need to figure out how next config maps onto this... (these are files that are copied as a separate process to next build and don't use node api's)

**Reply by chrisui** [reply]
furthermore: https://github.com/vercel/next.js/discussions/14807

**Comment by betocode** [comment]
bump, i got the same issue, i use process.cwd() to access a file(inside nextjs project) which work ok locally but not on vercel

**Comment by wh1t3h47** [comment]
Isn't `proccess.cwd()` totally unreliable? It may or may not point to the application directory, if I start the application from another directory, it will point to this other directory, like in a monorepo. Is there any workaround to it? module.path doesn't work, module.paths doesn't as well

**Reply by GabenGar** [reply]
Save the path to your application directory relative to `process.cwd()` in a variable and use it in your path calls.
